### ISSUE TITLE
It is not possible to insert a hyperlink like mailto:email@address.com into a document

### ISSUE THREAD

**Issue by bkeevil** [CollaboraOnline/online#1371]
**Describe the bug**
It is not possible to insert a hyperlink like mailto:email@address.com into a word processor document (and possibly other document types).

Collabora adds "http://" to the start of everything, without checking if it is a mailto: or other type of url first.

Collabora Version: 6.4-19

**Comment by cornouws** [comment]
I confirm the issue.
In Online, Insert > Hyperlink shows a small specific dialog, not the full dialog from Desktop, where one can make a choice for the type of hyperlink.

**Comment by cornouws** [comment]
But it is more a RFE, than a bug, I would say.

**Comment by bkeevil** [comment]
It should check if the text starts with "mailto:" and not add "http://" if it does.

**Comment by pedropintosilva** [comment]
What do you think about this @eszkadev ? or should we convert this into an easy hack?

**Comment by eszkadev** [comment]
@pedropintosilva yes, here is pointer: https://github.com/CollaboraOnline/online/blob/master/browser/src/control/Toolbar.js#L717

**Comment by Netzvamp** [comment]
https://github.com/CollaboraOnline/online/blob/e6a06eae1cc0495940d9870cc0718cd93676bdbe/browser/src/control/Toolbar.js#L740
A quick working hack:
```javascript
makeURLFromStr: function(str) {
		if (!(str.toLowerCase().startsWith('http://') || str.toLowerCase().startsWith('https://') || str.toLowerCase().startsWith('mailto:') || str.toLowerCase().startsWith('tel:'))) {
			str = 'http://' + str;
		}
		return str;
	},
```

or a bit more nicer:

```javascript
makeURLFromStr: function(str) {
		var protocols = ['http://', 'https://', 'mailto:', 'tel:'];
		var lowerStr = str.toLowerCase();
		var hasProtocol = false;
		for (var i = 0; i < protocols.length; i++) {
			if (lowerStr.startsWith(protocols[i])) {
				hasProtocol = true;
				break;
			}
		}
		if (!hasProtocol) {
			str = 'http://' + str;
		}
		return str;
	},
```

**Comment by mdoggydog** [comment]
I opened #7917 a couple of days ago, without realizing this issue was already open here.

The title of this issue here does not quite do justice to the nature of the issue:  the problem is not just that adding a new `mailto:` hyperlink is impossible; the problem is that CollaboraOnline insists that every link it touches starts with `http://` or `https://`.  Aside from making it impossible to add any new non-HTTP link, it also makes it easy to accidentally screw up existing non-HTTP links, without noticing the change and without being able to fix it without exporting the file and using another tool.

As I opined in #7917, I think the proper fix at this point is very simple:  do not attempt to do _any_ syntax checking on the URL provided by the user...

 > Even if you checked for any syntactically valid scheme component (via a regexp, e.g., `[a-zA-Z][a-zA-Z0-9+.-]*:`), you cannot predict what the user is trying to accomplish with the link (perhaps a missing scheme is _intentional_).

The best option is to just accept the user's input at face value.  This is also very easy to implement, and even easier to maintain.  :smiley: 

(If CollaboraOnline introduces a more elaborate hyperlink dialog in the future, like desktop LibreOffice has, with templates/forms to assist the user in constructing different types of URL's, then it will have a basis for understanding the user intent and performing syntax checking.  But, even then, it should allow to user to enter a non-conforming URL if that is what the user wants --- and it should preserve non-conforming URL's which already exist in a document.)

**Comment by Faibk** [comment]
At the very least there should be a way of disabling this behaviour or editing a hyperlink after inserting. 
However, I agree that this is a typical case of a "feature" that does more harm than good and could be dropped entirely. HTTP is not even considered a safe default anymore, it should be HTTPS.

I don't think this issue should be considered just an enhancement but a bug, since it is not only inconsistent with the Desktop version but corrupts documents created with it (editing a non-http link created in Desktop breaks it with no way of fixing).

**Comment by Darshan-upadhyay1110** [comment]
Fixed by commit [7f0c21410ef7](https://github.com/CollaboraOnline/online/pull/9157) 'Toolbar: fix hyperlink protocol preservation' (May 2024): makeURLFromStr now preserves mailto:, ftp:, http(s): prefixes and automatically prepends mailto: to bare email addresses. Inserting a mailto: hyperlink works as requested. tel: links (raised later in the thread) could be a small follow-up issue if still wanted.
